src/Semaphore.ts

```diff
@@ -0,0 +1,46 @@
+type Unlock = () => void;
+
+export class Semaphore {
+
+  private _maxAllowed: number;
+  private _permits: number;
+  private _queue: Unlock[] = [];
+
+  constructor(maxAllowed: number = 1) {
+    this._maxAllowed = maxAllowed;
+    this._permits = maxAllowed;
+  }
+
+  get permits() {
+    return this._permits;
+  }
+
+  get queueLength() {
+    return this._queue.length;
+  }
+
+  get acquired() {
+    return this._maxAllowed - this._permits;
+  }
+
+  public async acquire(): Promise<void> {
+    if (this._permits > 0) {
+      this._permits -= 1;
+      return;
+    }
+
+    return new Promise( resolve => this._queue.push(resolve) );
+  }
+
+  public release(): void {
+    if (this._queue.length) {
+      this._queue.shift()?.();
+    } else {
+      this._permits += 1;
+    }
+
+    if (this._permits > this._maxAllowed) {
+      console.error('Release() has been called without previous acquire!');
+    }
+  }
+}
```

Comment:
This code implements a Semaphore, which is a synchronization primitive that can be used to control access to a common resource by multiple processes in a concurrent system such as a multi-threaded program.

Here are some observations and suggestions:

1. **Constructor Default Parameter**: The constructor allows for an optional `maxAllowed` parameter with default value 1, which is good because it makes the semaphore behave like a lock (binary semaphore) when no argument is passed.

2. **permit Getter**: The getter for `permits` provides read access to the current number of available permits, which is useful for debugging or monitoring purposes.

3. **queueLength Getter**: Similarly, exposing the queue length with a getter allows external monitoring without giving direct access to the queue.

4. **acquired Getter**: This computed property is helpful for understanding how many permits have been acquired at any given time.

5. **Acquire Function**:
    - The `acquire()` function correctly decrements `_permits` if permits are available. 
    - It returns a promise that resolves once the semaphore allows acquiring, which lets it work in an asynchronous environment.
    - However, there is no handling for race conditions. There might be cases when multiple calls to `acquire()` see that `this._permits > 0` and they all proceed, leading to more than the allowed number of permits being acquired before `_permits` is updated.

6. **Release Function**:
    - It prioritizes releasing waiting tasks (`_queue.shift()?.()`) before incrementing the count of `_permits`, which is good practice.
    - A potential issue here is with the error checking: after calling the resolve function from the queue, other code could still call release. To prevent this, you should ideally re-check the condition after modifying the queue but before returning.
    - The check for an incorrect release (i.e., releasing more permits than the maximum) is done after incrementing `_permits`. This allows the `_permits` to go above `_maxAllowed` before catching the mistake, which could lead to bugs. This check should be upstream, and it may be beneficial to throw an exception instead of just logging an error.

7. **Thread Safety / Concurrency Considerations**:
    - Since JavaScript's event loop model executes all code on a single thread, true parallelism is not possible, which means the risk of race conditions due to concurrency in typical JavaScript runtime environments (like browsers or Node.js) is low. Still, this code does not inherently protect against such issues if used in an environment that enables true concurrency (like with Web Workers), or when using platforms that do enable multi-threaded JavaScript.

8. **Improvements**:
    - For better encapsulation and readability, you might want to use TypeScript's private keyword for `_maxAllowed`, `_permits`, and `_queue`, assuming they aren't meant to be accessible outside of the class.
    - Consider adding type annotations to class methods for parameters and return types even if they're inferred for enhanced readability and maintainability.
    - Think about implementing a timeout feature for the `acquire()` method, allowing a wait to fail after a certain period.
    - The semaphore implementation lacks disposal logic; it might be helpful to add a method to clear the queue and release all resources, especially for long-running applications.

9. **Error Handling**: The error message in the release method should be improved to provide more context, such as how many extra releases were attempted.

Overall, while the basic mechanism of theSemaphore class is implemented correctly, careful attention should be paid to the concurrency aspects, including additional checks and potentially making the data access atomic to avoid race conditions in certain environments or scenarios.